examples/pytorch/ogb/directional_GSN/main.py

```diff
@@ -0,0 +1,638 @@
+        super().__init__()
+        list_FC_layers = [nn.Linear(input_dim // 2 ** l, input_dim // 2 ** (l + 1), bias=True) for l in range(L)]
+        list_FC_layers.append(nn.Linear(input_dim // 2 ** L, output_dim, bias=True))
+        self.FC_layers = nn.ModuleList(list_FC_layers)
+        self.L = L
+
+    def forward(self, x):
+        y = x
+        for l in range(self.L):
+            y = self.FC_layers[l](y)
+            y = F.relu(y)
+        y = self.FC_layers[self.L](y)
+        return y
+
+class DGNNet(nn.Module):
+    def __init__(self, net_params):
+        super().__init__()
+        hidden_dim = net_params['hidden_dim']
+        out_dim = net_params['out_dim']
+        in_feat_dropout = net_params['in_feat_dropout']
+        dropout = net_params['dropout']
+        n_layers = net_params['L']
+        self.graph_norm = net_params['graph_norm']
+        self.batch_norm = net_params['batch_norm']
+        self.aggregators = net_params['aggregators']
+        self.avg_d = net_params['avg_d']
+        self.residual = net_params['residual']
+        device = net_params['device']
+        self.device = device
+
+        self.in_feat_dropout = nn.Dropout(in_feat_dropout)
+
+        self.embedding_h = AtomEncoder(emb_dim=hidden_dim)
+
+        # retrieve the aggregators and scalers functions
+        self.aggregators = [AGGREGATORS[aggr] for aggr in self.aggregators.split()]
+
+        self.layers = nn.ModuleList([DGNLayer(in_dim=hidden_dim, out_dim=hidden_dim, dropout=dropout, graph_norm=self.graph_norm,
+                      batch_norm=self.batch_norm, residual=self.residual, aggregators=self.aggregators,
+                      avg_d=self.avg_d) for _ in range(n_layers - 1)])
+        self.layers.append(DGNLayer(in_dim=hidden_dim, out_dim=out_dim, dropout=dropout,
+                                    graph_norm=self.graph_norm, batch_norm=self.batch_norm,
+                                    residual=self.residual, aggregators=self.aggregators, 
+                                    avg_d=self.avg_d))
+
+        # 128 out dim since ogbg-molpcba has 128 tasks
+        self.MLP_layer = MLPReadout(out_dim, 128)
+
+    def forward(self, g, h, e, snorm_n):
+        h = self.embedding_h(h)
+        h = self.in_feat_dropout(h)
+
+        for i, conv in enumerate(self.layers):
+            h_t = conv(g, h, e, snorm_n)
+            h = h_t
+
+        g.ndata['h'] = h
+
+        hg = dgl.mean_nodes(g, 'h')
+
+        return self.MLP_layer(hg)
+
+    def loss(self, scores, labels):
+        
+        loss = 0
+        loss_fn = torch.nn.BCEWithLogitsLoss()
+
+        for i in range(labels.shape[1]):
+            # ignore nan values
+            is_labeled = labels[:,i] == labels[:,i]
+
+            # otherwise loss will be 'nan'
+            if (len(labels[is_labeled, i]) == 0):
+                continue
+
+            loss += loss_fn(scores[is_labeled,i], labels[is_labeled,i].to('cuda'))
+        return loss
```

Comment:
Perhaps you can first calculate the loss matrix, multiply it by the `is_labeled` matrix and then take the sum or mean.